capture/lib/ot_communication.py

```diff
@@ -404,62 +404,153 @@ def read_ciphertext(self, len_bytes):
                         return batch_digest[0:len_bytes]
                     except Exception:
                         pass  # noqa: E302
-        return self.target.simpleserial_read("r", len_bytes, ack=False)
 
 
 class OTSHA3:
-    def __init__(self, target) -> None:
+    def __init__(self, target, protocol: str,
+                 port: Optional[OTUART] = None) -> None:
         self.target = target
+        self.simple_serial = True
+        self.port = port
+        if protocol == "ujson":
+            self.simple_serial = False
+            # Init the SHA3 core.
+            self.port.write(json.dumps("ShaSca").encode("ascii"))
+            self.port.write(json.dumps("Init").encode("ascii"))
+
+    def _ujson_sha_sca_cmd(self):
+        # TODO: without the delay, the device uJSON command handler program
+        # does not recognize the commands.
+        time.sleep(0.01)
+        self.port.write(json.dumps("ShaSca").encode("ascii"))
+
+    def _ujson_sha_sca_ack(self):
+        # Wait for ack.
+        while True:
+            read_line = str(self.port.readline())
+            if "RESP_OK" in read_line:
+                json_string = read_line.split("RESP_OK:")[1].split(" CRC:")[0]
+                try:
+                    status = json.loads(json_string)["status"]
+                    if status[0] != 0:
+                        raise Exception("Batch mode acknowledge error: Device and host not in sync")
+                    return status
+                except Exception:
+                    raise Exception("Batch mode acknowledge error: Device and host not in sync")
+            else:
+                raise Exception("Batch mode acknowledge error: Device and host not in sync")
```

Comment:
There are two commands where `_ujson_sha_sca_ack()`  is called: `set_mask_off/on()` and `absorb_batch()`.
Thus, the error message might be missleading if this error happens in non-batch mode for the mask on/off function.
I think it is sufficient to just say "Acknowledge error:..." and use the exception information to track the function where it occurred.

Reply:
Good point, I've changed the message.